### PR TITLE
[Consensus] `VoteAggregator` pruning revision

### PR DESCRIPTION
**Pull request by durkmurder**
### Context

This PR implements requested changes for pruning `VoteAggregator` and handling of finalization events for compliance engine.

### CODE REVIEW

File: consensus/hotstuff/voteaggregator/vote_aggregator.go [hunk 1]
```diff
@@ -28,12 +28,14 @@ const defaultVoteQueueCapacity = 1000
...
-	collectors          hotstuff.VoteCollectors
-	queuedVotesNotifier engine.Notifier
-	queuedVotes         *fifoqueue.FifoQueue
+	log                        zerolog.Logger
+	notifier                   hotstuff.Consumer
+	lowestRetainedView         counters.StrictMonotonousCounter // lowest view, for which we still process votes
+	collectors                 hotstuff.VoteCollectors
+	queuedVotesNotifier        engine.Notifier
+	finalizationEventsNotifier engine.Notifier
+	finalizedView              counters.StrictMonotonousCounter
```
Comment (zhangchiqing) [inline consensus/hotstuff/voteaggregator/vote_aggregator.go:37]:
```suggestion
	finalizedView              counters.StrictMonotonousCounter // cache the last finalized view to queue up the pruning work, and unblock the caller who's delivering the finalization event.
```

**Review by zhangchiqing — APPROVED** [review]

File: engine/collection/compliance/engine_test.go [hunk 2]
```diff
@@ -319,15 +319,14 @@ func (cs *ComplianceSuite) TestOnFinalizedBlock() {
...
 		done.Toggle()
 	}).Return(nil).Once()
-	cs.engine.OnFinalizedBlock(finalizedBlock.ID())
-
-	// matching engine has at least 100ms ticks for processing events
-	time.Sleep(1 * time.Second)
+	cs.pending.On("Size").Return(uint(0)).Once()
+	cs.engine.OnFinalizedBlock(model.BlockFromFlow(finalizedBlock.Header, finalizedBlock.Header.View-1))
 
 	require.Eventually(cs.T(), done.Load, time.Second, time.Millisecond*20)
```
Comment (AlexHentschel) [inline engine/collection/compliance/engine_test.go:329]: Same here: would suggest to leave concurrency up to testify and just write:
```golang
func (cs *ComplianceSuite) TestOnFinalizedBlock() {
	...

	*cs.pending = module.PendingClusterBlockBuffer{}
	cs.pending.On("PruneByView", finalizedBlock.Header.View).Once()
	cs.pending.On("Size").Return(uint(0)).Once()
	cs.engine.OnFinalizedBlock(model.BlockFromFlow(finalizedBlock.Header, finalizedBlock.Header.View-1))

	require.Eventually(cs.T(), func() bool {
		return cs.pending.AssertCalled(cs.T(), "PruneByView", finalizedBlock.Header.View)
	}, time.Second, time.Millisecond*20)
	cs.pending.AssertExpectations(cs.T())
}
```

File: engine/consensus/compliance/engine_test.go [hunk 3]
```diff
@@ -219,15 +220,14 @@ func (cs *ComplianceSuite) TestOnFinalizedBlock() {
 	cs.head = &finalizedBlock
 
 	done := atomic.NewBool(false)
```
Comment (AlexHentschel) [inline engine/consensus/compliance/engine_test.go:222]: same as [my comment above](url): would suggest to leaven concurrency handling to testify

File: module/buffer.go [hunk 4]
```diff
@@ -19,7 +19,7 @@ type PendingBlockBuffer interface {
 
 	DropForParent(parentID flow.Identifier)
 
-	PruneByHeight(height uint64)
+	PruneByView(view uint64)
```
Comment (AlexHentschel) [inline module/buffer.go:22]: brief goDoc would be great specifically stating whether `view` is included in pruning or not:
```suggestion
	// PruneByView prunes any pending blocks with views less or equal to the given view.
	PruneByView(view uint64)
```

File: module/buffer.go [hunk 5]
```diff
@@ -35,7 +35,7 @@ type PendingClusterBlockBuffer interface {
...
```
Comment (AlexHentschel) [inline module/buffer.go:38]:
```suggestion
	// PruneByView prunes any pending cluster blocks with views less or equal to the given view.
	PruneByView(view uint64)
```

File: module/buffer/pending_blocks.go [hunk 6]
```diff
@@ -59,8 +59,8 @@ func (b *PendingBlocks) DropForParent(parentID flow.Identifier) {
 	b.backend.dropForParent(parentID)
 }
 
-func (b *PendingBlocks) PruneByHeight(height uint64) {
-	b.backend.pruneByHeight(height)
+func (b *PendingBlocks) PruneByView(view uint64) {
```
Comment (AlexHentschel) [inline module/buffer/pending_blocks.go:62]:
```suggestion
// PruneByView prunes any pending blocks with views less or equal to the given view.
func (b *PendingBlocks) PruneByView(view uint64) {
```

File: module/buffer/pending_cluster_blocks.go [hunk 7]
```diff
@@ -57,8 +57,8 @@ func (b *PendingClusterBlocks) DropForParent(parentID flow.Identifier) {
 	b.backend.dropForParent(parentID)
 }
 
-func (b *PendingClusterBlocks) PruneByHeight(height uint64) {
-	b.backend.pruneByHeight(height)
+func (b *PendingClusterBlocks) PruneByView(view uint64) {
```
Comment (AlexHentschel) [inline module/buffer/pending_cluster_blocks.go:60]:
```suggestion
// PruneByView prunes any pending cluster blocks with views less or equal to the given view.
func (b *PendingClusterBlocks) PruneByView(view uint64) {
```

**Review by AlexHentschel — APPROVED** [review]
Very clean code. ✨  Thanks. 👏  
* Had a few stylistic comments. 
* Only other this is that I randomly noticed a bug in our [`PendingBlocks.Size()`](https://github.com/onflow/flow-go/blob/0d5c07670f6f7bf61c6e7d5aac5e02fc61ddbba3/module/buffer/pending_blocks.go#L66-L68) and [`PendingClusterBlocks.Size()`](https://github.com/onflow/flow-go/blob/0d5c07670f6f7bf61c6e7d5aac5e02fc61ddbba3/module/buffer/pending_cluster_blocks.go#L64-L66) methods, which we use for metrics:
   * They directly reads the length of the backend map **without any concurrency protection**. For all other methods, the backend handles the concurrency ... except here. 
   * would suggest to include a concurrency-safe method into the backend
      ```golang
      // size returns the number of elements stored in teh backend
      func (b *backend) size() uint {
      	b.mu.RLock()
      	defer b.mu.RUnlock()
      	return uint(len(b.blocksByID))
      }
      ```
      which `PendingBlocks.Size()` and `PendingClusterBlocks.Size()` just call

File: engine/collection/compliance/core.go [hunk 8]
```diff
@@ -314,20 +314,18 @@ func (c *Core) prunePendingCache() {
 	}
 
 	// remove all pending blocks at or below the finalized height
-	c.pending.PruneByHeight(final.Height)
+	c.pending.PruneByView(final.Height)
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply engine/collection/compliance/core.go:317]: Forgot to remove

**Pull request merged**